### PR TITLE
[dv, Csrng] Add chip_sw_csrng_fuse_en_sw_app_read test

### PR DESCRIPTION
**Pull request by engdoreis**
Add `chip_sw_csrng_fuse_en_sw_app_read` chip level test:
 - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit set to enabled in the `uvm_test_seq`.
- Issue an instantiate command to request entropy.
- Verify that SW can read the internal states.
- Reset the chip and repeat the steps above, but this time, with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit set to 0.
- Verify that the SW reads back all zeros when reading the internal states.

Fixes: #13221

### CODE REVIEW

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+ * y = OTP_CTRL_PARAM_HW_CFG_OFFSET (1664)
+ * IFETCH_OFFSET = (x - y) = 64
+ */
+static const uint32_t kOtpIfetchHwRelativeOffset =
+    OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET - OTP_CTRL_PARAM_HW_CFG_OFFSET;
+
+static const uint32_t kOtpCsrngFwReadOffset =
+    (OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET -
+     OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET) *
+    8;
```
Comment (cindychip) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:35]: Could you briefly add a comment to explain why the `kOtpCsrngFwReadOffset` needs to multiple by 8 but `kOtpIfetchHwRelativeOffset` does not?

**Review by cindychip — COMMENTED** [review]
I am not very familiar with CSRNG but left a comment regarding OTP part ;)

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    LOG_INFO("Powered up for the first time");
+    check_csrng_fuse_enabled(true);
+    test_fuse_enable(&csrng);
+
+    // Reboot device.
+    rstmgr_testutils_reason_clear();
+    CHECK_DIF_OK(dif_rstmgr_software_device_reset(&rstmgr));
+    // This log message is extremely important for the test, as the
+    // `uvm_test_seq` uses it to change the otp values.
+    LOG_INFO("Software reseting!");
```
Comment (matutem) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:209]: Nit: ... resetting

**Review by matutem — COMMENTED** [review]
Minor nit, but I defer to others with more expertise in this area.

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 1 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:40]: Done, I hope the comment is clear enough. Please let me know if not.

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 2 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c]: Good catch

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv [hunk 3]
```diff
@@ -28,11 +28,11 @@ class chip_sw_entropy_src_fuse_vseq extends chip_sw_base_vseq;
     super.body();
 
     forever begin
-      `DV_WAIT(cfg.sw_logger_vif.printed_log == "Software reseting!",
+      `DV_WAIT(cfg.sw_logger_vif.printed_log == "Software resetting!",
```
Comment (marnovandermaas) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv:31]: Nice catch!

File: hw/top_earlgrey/data/chip_testplan.hjson [hunk 4]
```diff
@@ -1967,10 +1967,12 @@
       name: chip_sw_csrng_fuse_en_sw_app_read
       desc: '''Verify the fuse input to CSRNG.
 
-            - Initialize the OTP with this fuse bit set to 1.
+            - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit 
+              set to enabled in the `uvm_test_seq`.
             - Issue an instantiate command to request entropy.
             - Verify that SW can read the internal states.
```
Comment (marnovandermaas) [inline hw/top_earlgrey/data/chip_testplan.hjson:1973]: Nit:
states -> state
or
states -> state variables

File: hw/top_earlgrey/data/chip_testplan.hjson [hunk 5]
```diff
@@ -1967,10 +1967,12 @@
...
-            - Reset the chip and repeat the steps above, but this time, with OTP fuse bit set to 0.
+            - Reset the chip and repeat the steps above, but this time, with 
+              `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit set to 0.
             - Verify that the SW reads back all zeros when reading the internal states.
```
Comment (marnovandermaas) [inline hw/top_earlgrey/data/chip_testplan.hjson:1976]: Nit: same as above

File: hw/top_earlgrey/data/chip_testplan.hjson [hunk 6]
```diff
@@ -1967,10 +1967,12 @@
       name: chip_sw_csrng_fuse_en_sw_app_read
       desc: '''Verify the fuse input to CSRNG.
 
-            - Initialize the OTP with this fuse bit set to 1.
+            - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit
```
Comment (marnovandermaas) [inline hw/top_earlgrey/data/chip_testplan.hjson]: It might be more useful to write a sentence about this fuse bit rather than use the name used in the code. For example:
the fuse bit that controls whether software can read the CSRNG state.

File: hw/top_earlgrey/data/chip_testplan.hjson [hunk 7]
```diff
@@ -1967,16 +1967,16 @@
       name: chip_sw_csrng_fuse_en_sw_app_read
       desc: '''Verify the fuse input to CSRNG.
 
-            - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit 
+            - Initialize the OTP with `HW_CFG.OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET` fuse bit
```
Comment (marnovandermaas) [inline hw/top_earlgrey/data/chip_testplan.hjson:1970]: Nit: squash this change into the previous commit.

File: hw/top_earlgrey/data/chip_testplan.hjson [hunk 8]
```diff
@@ -1967,16 +1967,16 @@
...
               set to enabled in the `uvm_test_seq`.
             - Issue an instantiate command to request entropy.
             - Verify that SW can read the internal states.
-            - Reset the chip and repeat the steps above, but this time, with 
+            - Reset the chip and repeat the steps above, but this time, with
```
Comment (marnovandermaas) [inline hw/top_earlgrey/data/chip_testplan.hjson:1974]: Nit: squash this change into the previous commit.

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv [hunk 9]
```diff
@@ -35,7 +35,7 @@ class chip_sw_entropy_src_fuse_vseq extends chip_sw_base_vseq;
       if (cfg.sw_logger_vif.printed_log == "Software resetting!") begin
         cfg.mem_bkdr_util_h[Otp].otp_write_hw_cfg_partition(
           .device_id(DEVICE_ID), .manuf_state(MANUF_STATE),
-          .en_sram_ifetch(MUBI8FALSE), .en_csrng_sw_app_read(MUBI8TRUE),
+          .en_sram_ifetch(MUBI8FALSE), .en_csrng_sw_app_read(MUBI8FALSE),
```
Comment (marnovandermaas) [inline hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv:38]: Why is this change necessary? Could you add a comment in the code on why you have chosen these initial states?

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+ * not aligned, we read from the previous aligned address and use the following
+ * offset in bits to access the `csrng_sw_app_read` byte.
+ */
+static const uint32_t kOtpCsrngFwReadBitOffset =
+    (OTP_CTRL_PARAM_EN_CSRNG_SW_APP_READ_OFFSET -
+     OTP_CTRL_PARAM_EN_SRAM_IFETCH_OFFSET) *
+    8;
+
+enum {
+  kExpectedOutputLen = 16,
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: Why is this an enum instead of const int?

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+  CHECK_DIF_OK(dif_otp_ctrl_dai_read32_end(&otp, &value));
+  multi_bit_bool_t enable = bitfield_field32_read(
+      value,
+      (bitfield_field32_t){.mask = 0xff, .index = kOtpCsrngFwReadBitOffset});
+  CHECK((enable == kMultiBitBool8True) == expected,
+        "`fw_enable` not expected (%x)", enable);
+}
+
+/**
+ * This test:
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:182]: Nit: This test -> This test takes the following steps:
(To be consistent with the verbs below.)

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+
+  dif_rstmgr_t rstmgr;
+  dif_rstmgr_reset_info_bitfield_t info;
+  CHECK_DIF_OK(dif_rstmgr_init(
+      mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+  info = rstmgr_testutils_reason_get();
+
+  if (info == kDifRstmgrResetInfoPor) {
+    LOG_INFO("Powered up for the first time");
+    check_csrng_fuse_enabled(true);
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:206]: You check whether the fuse is true here, but where did you set it to begin with?

Comment (marnovandermaas) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:206]: If the default is true, then how is it unset for the else case below?

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+                                     &kEntropyInput));
+
+  const dif_csrng_internal_state_t kExpectedState = {
+      .reseed_counter = 1,
+      .v = {0x06b8f59e, 0x43c0b2c2, 0x21052502, 0x217b5214},
+      .key = {0x941709fd, 0xd8a25860, 0x861aecf3, 0x98a701a1, 0x0eb2c33b,
+              0x74c08fad, 0x632d5227, 0x8c52f901},
+      .instantiated = true,
+      .fips_compliance = false,
+  };
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:99]: How did you generate the input and expected state? Can you put a reference in a comment?

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+  check_internal_state(csrng, &kExpectedState);
+
+  // TODO(#13342): csrng does not provide a linear output order. For example,
+  // note the test vector output word order: 12,13,14,15 8,9,10,11 4,5,6,7
+  // 0,1,2,3.
+  const uint32_t kExpectedOutput[kExpectedOutputLen] = {
+      0xe48bb8cb, 0x1012c84c, 0x5af8a7f1, 0xd1c07cd9, 0xdf82ab22, 0x771c619b,
+      0xd40fccb1, 0x87189e99, 0x510494b3, 0x64f7ac0c, 0x2581f391, 0x80b1dc2f,
+      0x793e01c5, 0x87b107ae, 0xdb17514c, 0xa43c41b7,
+  };
```
Comment (marnovandermaas) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:134]: As above, where do these values come from, and can we add a reference to where to find them or how to generate them?

**Review by marnovandermaas — COMMENTED** [review]
Nice work, I just have a few questions, mostly to improve my understanding.

**Review by moidx — APPROVED** [review]
Hi @engdoreis, since we have been using the csrng kat test in various test cases, it probably makes sense to abstract it into a library to reduce the amount of duplicate code. Thanks!

File: hw/top_earlgrey/data/chip_testplan.hjson (hunk 6 above)
Comment (engdoreis) [reply hw/top_earlgrey/data/chip_testplan.hjson]: done

File: hw/top_earlgrey/data/chip_testplan.hjson (hunk 4 above)
Comment (engdoreis) [reply hw/top_earlgrey/data/chip_testplan.hjson:1973]: done

File: hw/top_earlgrey/data/chip_testplan.hjson (hunk 5 above)
Comment (engdoreis) [reply hw/top_earlgrey/data/chip_testplan.hjson:1976]: done

File: hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv (hunk 9 above)
Comment (engdoreis) [reply hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv:38]: Comment added. It is basically documented in the test plan.

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 10 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: It's broadly used on the opentitan and I'm not sure the main reason. I guess it has better type safity features.

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 11 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:182]: done

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 12 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:206]: The OTP can not be set by SW, so it is  done in the dv seq file (hw/top_earlgrey/dv/env/seq_lib/chip_sw_entropy_src_fuse_vseq.sv).
This is documented before the main function.

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 13 above)
Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:99]: This is a copy of another test, which I would be moving to a test utils as recomented by @moidx.
@moidx Would you have the references for this values?

Comment (moidx) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:99]: Hi @engdoreis, you can probably copy this docstring to the new location. It documents the data source for the test vectors. You can also tag a `// TODO(#6643):` for a follow up cleanup, and to address @marnovandermaas's comment.

cross-referencing: #6643.

https://github.com/lowRISC/opentitan/blob/42009f4a57e0e60ae808311485d15dbdc04aa1e6/sw/device/tests/csrng_kat_test.c#L106-L136

Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:99]: Thanks @moidx

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c (hunk 10 above)
Comment (moidx) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: @alphan can provide more details here.

Should we update the style guide to capture the rationale?

Comment (alphan) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: Sure! In a nutshell, We use `enum`s to define _named_ constants that
* appear in the symbol table for easier debugging and don't evaporate like macros do, and
* don't use any memory.

If the constant is defined merely to publish _some_ value, we don't name the type. If, on the other hand, the `enum` has some significance for the API, we name the type (using `typedef`).

`int` is not really an option because: it cannot be used as `case` labels in `switch` statements, and it cannot be used as the dimension of (global) arrays. `int` could be inlined, but depends on the compiler and optimization settings. Arguably, the only advantage of an `int` is you can take its address -- but you can always define one and initialize with an `enum` if needed.

One exception is, if a constant will be used in `asm`, then `#define` is the only option.

See this [SO answer](https://stackoverflow.com/questions/1674032/static-const-vs-define-vs-enum) for more details.

P.S. If you want to try this, please make sure to pass `-Wall -Wextra -Werror -pedantic-errors`.

Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: Thanks @alphan, this was very enlightening for me!

Comment (alphan) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:43]: glad that I could help!

File: sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+    LOG_INFO("Powered up for the first time");
+    check_csrng_fuse_enabled(true);
+    test_fuse_enable(&csrng);
+
+    // Reboot device.
+    rstmgr_testutils_reason_clear();
+    CHECK_DIF_OK(dif_rstmgr_software_device_reset(&rstmgr));
+    // This log message is extremely important for the test, as the
+    // `uvm_test_seq` uses it to change the otp values.
+    LOG_INFO("Software resetting!");
```
Comment (sriyerg) [inline sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:155]: OR, the `chip_sw_entropy_src_fuse_vseq` could use the internal reset monitor to know a mid-test reset occurred to update the OTP region.

Comment (sriyerg) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:155]: https://github.com/lowRISC/opentitan/blob/edc3accc977ff4c57703f36460bb60c0e4cd02bd/hw/top_earlgrey/dv/env/chip_env_cfg.sv#L26

Comment (engdoreis) [reply sw/device/tests/sim_dv/csrng_fuse_en_sw_app_read.c:155]: Looks good, But it is better to open an issue for that, since there's two tests that use the current mechanism
